packages/core/src/actions/execute.ts

```diff
@@ -58,10 +58,90 @@ export const executeTask = async (args: {
       throw new Error(`Another executor has already claimed the bundle.`)
     }
 
-    logger.info(`Setting the state variables...`)
+    // We execute all actions in batches to reduce the total number of transactions and reduce the
+    // cost of a deployment in general. Approaching the maximum block gas limit can cause
+    // transactions to be executed slowly as a result of the algorithms that miners use to select
+    // which transactions to include. As a result, we restrict our total gas usage to a fraction of
+    // the block gas limit.
+    const latestBlock = await executor.provider.getBlock('latest')
+    const gasFraction = 2
+    const maxGasLimit = latestBlock.gasLimit.div(gasFraction)
+
+    /**
+     * Helper function for finding the maximum number of batch elements that can be executed from a
+     * given input list of actions. This is done by performing a binary search over the possible
+     * batch sizes and finding the largest batch size that does not exceed the maximum gas limit.
+     *
+     * @param actions List of actions to execute.
+     * @returns Maximum number of actions that can be executed.
+     */
+    const findMaxBatchSize = async (actions: any[]): Promise<number> => {
+      let min = 0
+      let max = actions.length
+      while (min < max) {
+        const mid = Math.floor((min + max) / 2)
+        try {
+          const estimate =
+            await chugSplashManager.estimateGas.executeMultipleActions(
+              actions.slice(0, mid).map((action) => action.action),
+              actions.slice(0, mid).map((action) => action.proof.actionIndex),
+              actions.slice(0, mid).map((action) => action.proof.siblings)
+            )
+
+          if (estimate.gt(maxGasLimit)) {
+            max = mid
+          } else {
+            min = mid + 1
+          }
+        } catch (err) {
+          max = mid
+        }
+      }
+
+      // No possible size works, this is a problem and should never happen.
+      if (min === 0) {
+        throw new Error(
+          'unable to find a batch size that does not exceed the block gas limit'
+        )
+      }
+
+      return min
+    }
 
-    // Execute actions that have not been executed yet.
-    let currActionsExecuted = bundleState.actionsExecuted.toNumber()
```

Comment:
I think removing `currActionsExecuted` will be problematic since it allows us to skip actions that have already been executed. For example, say the executor gets halfway through a deployment then a transaction fails for some reason. Previously, the executor would be able to resume where it left off because of `currActionsExecuted`, but now it'll attempt to execute all of the actions again, including those that have previously been executed.